pkg/pgmodel/ingestor/trace/trace_batcher.go

```diff
@@ -0,0 +1,266 @@
+		trace.WithLinks(trace.Link{SpanContext: in.spanCtx}))
+	defer addReqSpan.End()
+	inSpans := in.payload.SpanCount()
+	if inSpans == 0 {
+		return
+	}
+	in.payload.ResourceSpans().MoveAndAppendTo(tb.traces.ResourceSpans())
+	tb.spanCount += inSpans
+	tb.reqStatus = append(tb.reqStatus, in.err)
+}
+
+func (tb *Batch) isFull() bool {
+	return tb.spanCount >= tb.maxSize
+}
+
+func (tb *Batch) isEmpty() bool {
+	return tb.spanCount == 0
+}
+
+// Batcher batches trace requests and sends batches to batch writer.
+// This is done to achieve better ingest performance especially b/c
+// Jaeger collector sends traces one by one.
+type Batcher struct {
+	in              chan insertTracesReq
+	stop            chan struct{}
+	batchWriter     *batchWriter
+	once            sync.Once
+	bufferedBatches chan Batch
+	wg              sync.WaitGroup
+	config          BatcherConfig
+}
+
+type BatcherConfig struct {
+	Batchers     int
+	Writers      int
+	MaxBatchSize int
+	BatchTimeout time.Duration
+}
+
+func NewBatcher(config BatcherConfig, writer Writer) *Batcher {
+	validateConfig(&config)
+	bufferedBatches := make(chan Batch, defaultMaxBufferedBatches)
+	return &Batcher{
+		batchWriter:     newBatchWriter(config.Writers, writer, bufferedBatches),
+		in:              make(chan insertTracesReq, defaultReqBufferSize),
+		stop:            make(chan struct{}),
+		bufferedBatches: bufferedBatches,
+		config:          config,
+	}
+}
+
+func validateConfig(config *BatcherConfig) {
+	if config.Batchers == 0 {
+		config.Batchers = DefaultBatchWorkers
+	}
+	if config.MaxBatchSize == 0 {
+		config.MaxBatchSize = DefaultBatchSize
+	}
+	if config.BatchTimeout == 0 {
+		config.BatchTimeout = DefaultBatchTimeout
+	}
+	if config.Batchers < 1 || config.Writers < 1 {
+		panic("number of batchers and writeres must be greater then zero")
+	}
+}
+
+func (b *Batcher) Run() {
+	for i := 0; i < b.config.Batchers; i++ {
+		b.wg.Add(1)
+		go func() {
+			defer b.wg.Done()
+			b.batch()
+		}()
+	}
+	b.batchWriter.run()
+}
+
+func (b *Batcher) batch() {
+	_, batcherSpan := tracer.Default().Start(context.Background(), "create-batches")
+	ticker := time.NewTicker(b.config.BatchTimeout)
+	defer ticker.Stop()
+	batch := NewBatch(b.config.MaxBatchSize)
+	batcherSpan.AddEvent("New Batch")
+	flushBatch := func(batch *Batch) *Batch {
+		b.bufferedBatches <- *batch
+		batcherSpan.AddEvent("Batch sent to buffer")
+		batcherSpan.AddEvent("New Batch")
+		metrics.IngestorPendingBatches.With(prometheus.Labels{"type": "trace"}).Inc()
+		return NewBatch(b.config.MaxBatchSize)
+	}
+	processReq := func(req insertTracesReq) {
+		metrics.IngestorRequestsQueued.With(prometheus.Labels{"type": "trace"}).Dec()
+		if batch.isFull() {
+			metrics.IngestorBatchFlushTotal.With(prometheus.Labels{"type": "trace", "reason": "size"}).Inc()
+			batch = flushBatch(batch)
+			ticker.Reset(b.config.BatchTimeout)
+		}
+		batch.add(req)
+	}
+	for {
+		select {
+		case item := <-b.in:
+			processReq(item)
+		case <-ticker.C:
```

Comment:
The ticker is only reset when we flush on size. Imho the reset on first request is matter of semantics - I feel that in practice it makes almost no difference.

Reply:
Not sure I understand why it would be reset on tick every tick?
Currently we only reset when we flush full batch.  To make any decent performance we need to flush by size - timeouts are just for worst case scenarios... Anyhow we can add reset on the first request in a batch to make sure that batch waits the timeout.